actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -548,63 +561,92 @@ def serializer
         def digest
           request.cookies_digest || "SHA1"
         end
-
-        def key_generator
-          request.key_generator
-        end
     end
 
-    class SignedCookieJar < AbstractCookieJar # :nodoc:
+    class SignedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       include SerializedCookieJars
 
       def initialize(parent_jar)
         super
-        secret = key_generator.generate_key(request.signed_cookie_salt)
-        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
+
+        secret = key_generator.generate_key(signed_cookie_salt)
+        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: signed_cookie_digest, serializer: SERIALIZER)
+
+        Array.wrap(signed_cookie_rotate_out).each do |outbound|
+          @verifier.rotate(*expand_outbound_rotation(outbound))
+        end
+
+        if rotate_out_legacy_signed_cookies?
+          @verifier.rotate request.secret_token, serializer: SERIALIZER
+        end
       end
 
       private
         def parse(name, signed_message)
-          deserialize name, @verifier.verified(signed_message)
+          deserialize(name, @verifier.verify(signed_message) { |message|
+            return deserialize(name, message).tap { |deserialized| self[name] = { value: deserialized } }
```

Comment:
The `verify` method (and the `decrypt_and_verify` method) call an optional block when a rotated instance is used to verify the message. We need to rewrite the cookie's value when this occurs and this optional block is utilized to signal that.

The tricky part is we need actually write the deserialize message so within this block `deserialize` is called and the cookie is updated. The `return` is there so the `parse` method returns the deserialized value from within the block. If we didn't return here, the outer `deserialize` would also get invoked and deserialization would effectively occur twice.

I do agree that `return` here is a little confusing. I'm definitely open to a better way of signal to the caller of `verify` or `decrypt_and_verify` when a rotated instance is used so that we can better handle this case.